nginx-ingress upgrade does not revert to previous state on failure

### What happened?

When upgrading from KKP2.18.4 to KKP2.19 using the `--migrate-upstream-nginx-ingress` flag, @dermorz and I noticed that when a helm release in the ingress-controller fails, its deployment is deleted and it will not be restored. This in turn blocks all future installs for the kkp-installer

### What should happen?

Install succeeds or at least the deleted deployment should be recovered from backup

### How to reproduce?

Run the installer, but have an invalid nginx values.yaml section (see example below)

```yaml
# example for broken section (podLabels should be keyvalue instead of list)
nginx:
    controller:
        podLabels:
            - "kkptest": itworks
            - "anotherone": yesyes
```

This will trigger our deployment deletion logic and afterwards expectedly fail when running the release. So far so good.

```
./kubermatic-installer219 deploy --config kubermatic-configuration.yaml --helm-values values.yaml --storageclass aws --migrate-upstream-nginx-ingress --migrate-upstream-cert-manager


INFO[10:07:12] 🛫 Initializing installer…                     edition="Enterprise Edition" version=v2.19.0-alpha.4
INFO[10:07:12] 🚦 Validating the provided configuration…
WARN[10:07:12]    Helm values: kubermaticOperator.imagePullSecret is empty, setting to spec.imagePullSecret from KubermaticConfiguration
INFO[10:07:12] ✅ Provided configuration is valid.
I0112 10:07:14.039016   34235 request.go:665] Waited for 1.040396968s due to client-side throttling, not priority and fairness, request: GET:https://pgdvnvrfnd.kubermatic.test-env-kkp.ee.lab.kubermatic.io:31305/apis/crd.projectcalico.org/v1?timeout=32s
INFO[10:07:14] 🧩 Deploying KKP master stack…
INFO[10:07:14]    💾 Deploying kubermatic-fast StorageClass…
INFO[10:07:14]    ✅ StorageClass exists, nothing to do.
INFO[10:07:14]    📦 Deploying nginx-ingress-controller…
INFO[10:07:14]       nginx-ingress-controller: 1.2.2 detected, performing upgrade to 1.3.4…
INFO[10:07:14]       Backing up old ingress deployment…
INFO[10:07:14]       attempting to store the deployment
INFO[10:07:14]       Deleting the deployment from the cluster
INFO[10:07:14]       Deploying Helm chart…
INFO[10:07:14]       Updating release from 1.2.2 to 1.3.4…
ERRO[10:07:18] ❌ Operation failed: failed to deploy nginx-ingress-controller: failed to deploy Helm release: failed to install: coalesce.go:199: warning: cannot overwrite table with non table for podLabels (map[])
coalesce.go:199: warning: cannot overwrite table with non table for podLabels (map[])
coalesce.go:199: warning: cannot overwrite table with non table for podLabels (map[])
Error: UPGRADE FAILED: template: nginx-ingress-controller/templates/service.yaml:26:22: executing "nginx-ingress-controller/templates/service.yaml" at <.Values.nginx.service.annotations>: nil pointer evaluating interface {}.annotations.

```

**However if you run it now with the correct values (see below), the installer will fail earlier because the deployment does not exist anymore (it got deleted and was never restored by the previous run)**

```yaml
nginx:
    controller:
        podLabels:
            "kkptest": itworks
            "anotherone": yesyes
```

```
❯ ./kubermatic-installer219 deploy --config kubermatic-configuration.yaml --helm-values values.yaml --storageclass aws --migrate-upstream-nginx-ingress --migrate-upstream-cert-manager

INFO[10:08:27] 🛫 Initializing installer…                     edition="Enterprise Edition" version=v2.19.0-alpha.4
INFO[10:08:27] 🚦 Validating the provided configuration…
WARN[10:08:27]    Helm values: kubermaticOperator.imagePullSecret is empty, setting to spec.imagePullSecret from KubermaticConfiguration
INFO[10:08:27] ✅ Provided configuration is valid.
I0112 10:08:28.360643   34836 request.go:665] Waited for 1.036590594s due to client-side throttling, not priority and fairness, request: GET:https://pgdvnvrfnd.kubermatic.test-env-kkp.ee.lab.kubermatic.io:31305/apis/acme.cert-manager.io/v1?timeout=32s
INFO[10:08:28] 🧩 Deploying KKP master stack…
INFO[10:08:28]    💾 Deploying kubermatic-fast StorageClass…
INFO[10:08:28]    ✅ StorageClass exists, nothing to do.
INFO[10:08:28]    📦 Deploying nginx-ingress-controller…
INFO[10:08:28]       nginx-ingress-controller: 1.2.2 detected, performing upgrade to 1.3.4…
INFO[10:08:28]       Backing up old ingress deployment…
INFO[10:08:28]       attempting to store the deployment
ERRO[10:08:28] ❌ Operation failed: failed to deploy nginx-ingress-controller: failed to upgrade nginx-ingress-controller: found more than one deployment (0) matching the nginx-ingress-controller release, stopping upgrade....
```

From the general installer user-experience we think, that either:
* we should roll-back the deployment on a failed helm release
* (maybe) we can change our strategy to first do the helm release and then just delete the pods to get them re-created
(only meant as possible suggestion, we are not too deep in the ideas that went into the implementation)


### Environment

- KKP Version: 2.18.4 to 2.19.alpha4
- Shared or separate master/seed clusters?: shared
- Master/seed cluster provider: AWS

--- a/pkg/install/stack/kubermatic-master/nginxingress.go
+++ b/pkg/install/stack/kubermatic-master/nginxingress.go
@@ -122,34 +122,33 @@ func upgradeNginxIngress(
 		Version: "v1",
 	})
 
-	if err := kubeClient.List(ctx, deploymentsList, client.InNamespace(NginxIngressControllerNamespace), client.MatchingLabels{
+	nginxMatcher := client.MatchingLabels{
 		"app.kubernetes.io/name":       "ingress-nginx",
 		"app.kubernetes.io/managed-by": "Helm",
 		"app.kubernetes.io/instance":   release.Name,
-	}); err != nil {
-		logger.Warn("Error querying API for the existing deployment, attempting to upgrade without removing it...")
-	} else {
-		// 2: store the deployment for backup
-		// There can be only one...
-		switch foundDeployments := len(deploymentsList.Items); foundDeployments {
-		case 1:
-			filename := fmt.Sprintf("backup_%s_%s.yaml", NginxIngressControllerReleaseName, backupTS)
-			logger.Infof("Attempting to store the deployment in file: %s", filename)
-			if err := util.DumpResources(ctx, filename, deploymentsList.Items); err != nil {
-				return fmt.Errorf("failed to back up the deployment, it is not removed: %v", err)
-			}
-
-			// 3: delete the deployment
-			logger.Info("Deleting the deployment from the cluster")
-			if err := kubeClient.Delete(ctx, &deploymentsList.Items[0]); err != nil {
-				return fmt.Errorf("failed to remove the deployment: %v\n\nuse backup file to check the changes and restore if needed", err)
-			}
-
-		case 0:
-			logger.Info("No deployments matching the criteria were found, assuming clean state and attempting to upgrade...")
-		default:
-			return fmt.Errorf("found more than one deployment (%d) matching the nginx-ingress-controller release, stopping upgrade - clean up manually before proceeding", len(deploymentsList.Items))
+	}
+	if err := kubeClient.List(ctx, deploymentsList, client.InNamespace(NginxIngressControllerNamespace), nginxMatcher); err != nil {
+		return fmt.Errorf("Error querying API for the existing Deployment object, aborting upgrade process.")
+	}
+	// 2: store the deployment for backup
+	// There can be only one...
+	switch foundDeployments := len(deploymentsList.Items); foundDeployments {
+	case 1:
+		filename := fmt.Sprintf("backup_%s_%s.yaml", NginxIngressControllerReleaseName, backupTS)
+		logger.Infof("Attempting to store the deployment in file: %s", filename)
+		if err := util.DumpResources(ctx, filename, deploymentsList.Items); err != nil {
+			return fmt.Errorf("failed to back up the deployment, it is not removed: %v", err)
 		}
+
+		// 3: delete the deployment
+		logger.Info("Deleting the deployment from the cluster")
+		if err := kubeClient.Delete(ctx, &deploymentsList.Items[0]); err != nil {
+			return fmt.Errorf("failed to remove the deployment: %v\n\nuse backup file to check the changes and restore if needed", err)
+		}
+	case 0:
+		logger.Info("No deployments matching the criteria were found, assuming clean state and attempting to upgrade...")
+	default:
+		return fmt.Errorf("found more than one deployment (%d) matching the nginx-ingress-controller release, stopping upgrade - clean up manually before proceeding", len(deploymentsList.Items))
 	}
 	return nil
 }

--- a/pkg/install/stack/kubermatic-master/nginxingress.go
+++ b/pkg/install/stack/kubermatic-master/nginxingress.go
@@ -62,6 +62,8 @@ func deployNginxIngressController(ctx context.Context, logger *logrus.Entry, kub
 	v13 := semver.MustParse("1.3.0")
 	backupTS := time.Now().Format("2006-01-02T150405")
 
+	isUpgrading := false
+
 	if release != nil && release.Version.LessThan(v13) && !chart.Version.LessThan(v13) {
 		if !opt.EnableNginxIngressMigration {
 			sublogger.Warn("To upgrade nginx-ingress-controller to a new version, the installer")
@@ -72,9 +74,10 @@ func deployNginxIngressController(ctx context.Context, logger *logrus.Entry, kub
 			return fmt.Errorf("user must acknowledge the migration using --migrate-upstream-nginx-ingress")
 		}
 
+		isUpgrading = true
 		err = upgradeNginxIngress(ctx, sublogger, kubeClient, helmClient, opt, chart, release, backupTS)
 		if err != nil {
-			return fmt.Errorf("failed to upgrade nginx-ingress-controller: %v", err)
+			return fmt.Errorf("failed to prepare nginx-ingress-controller for upgrade: %v", err)
 		}
 	}
 
@@ -83,6 +86,9 @@ func deployNginxIngressController(ctx context.Context, logger *logrus.Entry, kub
 	sublogger.Info("Deploying Helm chart…")
 
 	if err := util.DeployHelmChart(ctx, sublogger, helmClient, chart, NginxIngressControllerNamespace, NginxIngressControllerReleaseName, opt.HelmValues, false, opt.ForceHelmReleaseUpgrade, release); err != nil {
+		if isUpgrading {
+			return fmt.Errorf("failed to deploy Helm release: %v\n\nuse backup file to restore the deployment or re-try the installation after fixing any errors.", err)
+		}
 		return fmt.Errorf("failed to deploy Helm release: %v", err)
 	}
 
@@ -123,23 +129,26 @@ func upgradeNginxIngress(
 	}); err != nil {
 		logger.Warn("Error querying API for the existing deployment, attempting to upgrade without removing it...")
 	} else {
-		logger.Info("attempting to store the deployment")
 		// 2: store the deployment for backup
 		// There can be only one...
-		if len(deploymentsList.Items) == 1 {
+		switch foundDeployments := len(deploymentsList.Items); foundDeployments {
+		case 1:
 			filename := fmt.Sprintf("backup_%s_%s.yaml", NginxIngressControllerReleaseName, backupTS)
+			logger.Infof("Attempting to store the deployment in file: %s", filename)
 			if err := util.DumpResources(ctx, filename, deploymentsList.Items); err != nil {
-				return fmt.Errorf("failed to back up the deployment: %v", err)
+				return fmt.Errorf("failed to back up the deployment, it is not removed: %v", err)
 			}
 
 			// 3: delete the deployment
 			logger.Info("Deleting the deployment from the cluster")
 			if err := kubeClient.Delete(ctx, &deploymentsList.Items[0]); err != nil {
-				return fmt.Errorf("failed to remove the deployment: %v\n\nuse backup file: %s to check the changes and restore if needed", err, filename)
+				return fmt.Errorf("failed to remove the deployment: %v\n\nuse backup file to check the changes and restore if needed", err)
 			}
 
-		} else {
-			return fmt.Errorf("found more than one deployment (%d) matching the nginx-ingress-controller release, stopping upgrade...", len(deploymentsList.Items))
+		case 0:
+			logger.Info("No deployments matching the criteria were found, assuming clean state and attempting to upgrade...")
+		default:
+			return fmt.Errorf("found more than one deployment (%d) matching the nginx-ingress-controller release, stopping upgrade - clean up manually before proceeding", len(deploymentsList.Items))
 		}
 	}
 	return nil

